### PR TITLE
added rate limits to each room by author group and author

### PR DESCRIPTION
**Pull request by FrederickPu**
added rate limits on the number of pending requests for on a room for each user and group

Note: still have not created a constant for the maximum number of pending requests (it's a hardcoded value).

### CODE REVIEW

**Comment by FrederickPu** [comment]
#148

**Comment by ggggg** [comment]
Ideally, we would make it into a per-room setting in the future. Unfortunately, room settings arent really a thing right now...

**Comment by ggggg** [comment]
The original issue does state that limits should be "per room." If you could make that change, it'd be great!

**Comment by FrederickPu** [comment]
> The original issue does state that limits should be "per room." If you could make that change, it'd be great!
Isn't it already doing that since the db request already only finds pending requests with the current room name?

**Comment by ggggg** [comment]
Sorry, I meant configured per room, as in, making an adjustable setting for each room

**Comment by FrederickPu** [comment]
so you mean change the db so that the admin can configure pr room limits?

File: backend/src/models/requestsModel.ts [hunk 1]
```diff
@@ -265,6 +265,35 @@ export default {
...
     request.approvers = request.approvers ?? [];
 
+    const room = await db.room.findUnique({ where : { roomName : request.roomName } })
+    const pendingUser = await db.request.findMany({
+      where: {
+        authorUtorid: user.utorid,
+        status: RequestStatus.pending,
+        roomName: request.roomName,
+      }
+    })
```
Comment (ggggg) [inline backend/src/models/requestsModel.ts:275]: This info can be fetched when fetching this user, try limiting the number of DB queries

File: backend/src/models/requestsModel.ts [hunk 2]
```diff
@@ -265,6 +265,35 @@ export default {
...
+      }
+    }
+
+    const pendingGroup = await db.request.findMany({
+      where: {
+        groupId: request.groupId,
+        status: RequestStatus.pending,
+        roomName: request.roomName,
+      }
+    })
```
Comment (ggggg) [inline backend/src/models/requestsModel.ts:289]: Same as the other comment

**Review by ggggg — CHANGES_REQUESTED** [review]

File: backend/src/models/requestsModel.ts [hunk 3]
```diff
@@ -265,6 +265,35 @@ export default {
...
 
+    const room = await db.room.findUnique({ where : { roomName : request.roomName } })
+    const pendingUser = await db.request.findMany({
+      where: {
+        authorUtorid: user.utorid,
+        status: RequestStatus.pending,
+        roomName: request.roomName,
+      }
+    })
+    if (pendingUser.length > (room?.capacity || 10)) {
```
Comment (ggggg) [inline backend/src/models/requestsModel.ts:276]:
```suggestion
    if (pendingUser.length > (room?.requestLimit || 10)) {
```

File: backend/src/models/requestsModel.ts [hunk 4]
```diff
@@ -265,6 +265,35 @@ export default {
...
+    }
+
+    const pendingGroup = await db.request.findMany({
+      where: {
+        groupId: request.groupId,
+        status: RequestStatus.pending,
+        roomName: request.roomName,
+      }
+    })
+    if (pendingGroup.length > (room?.capacity || 10)) {
```
Comment (ggggg) [inline backend/src/models/requestsModel.ts:290]:
```suggestion
    if (pendingGroup.length > (room?.requestLimit || 10)) {
```

File: backend/src/models/requestsModel.ts [hunk 5]
```diff
@@ -265,6 +267,36 @@ export default {
...
+      return { status: 404, message: 'Group not found.'};
+    }
+
+    const room = await db.room.findUnique({ where : { roomName : request.roomName } })
+    if (!room) {
+      return { status: 404, message: 'Room not found.'};
+    }
+    if (userFetched.requests.length > room.requestLimit) {
+      return {
+        status: 403,
```
Comment (logonoff) [inline backend/src/models/requestsModel.ts:288]: should this be http 429

File: backend/src/models/requestsModel.ts [hunk 6]
```diff
@@ -265,6 +267,36 @@ export default {
...
+        message: 'User has too many pending requests.',
+      }
+    }
+
+    if (groupFetched.requests.length > room.requestLimit) {
+      return {
+        status: 403,
```
Comment (logonoff) [inline backend/src/models/requestsModel.ts:295]: should this be http 429: too many requests

File: backend/src/models/requestsModel.ts [hunk 7]
```diff
@@ -247,24 +247,56 @@ export default {
...
+        message: 'User has too many pending requests.',
+      }
+    }
+    if (userFetched.groups[0].requests.length >= room.requestLimit) {
+      return {
+        status: 429,
+        message: 'Group has too many pending requests.',
+      }
+    }
+
```
Comment (ggggg) [inline backend/src/models/requestsModel.ts:299]:
```suggestion
    if (userFetched.requests.length >= room.requestLimit || userFetched.groups[0].requests.length >= room.requestLimit) {
      return {
        status: 429,
        message: 'User has too many pending requests.',
      }
    }
   
```

**Review by ggggg — APPROVED** [review]

**Pull request merged**